svcb.go

```diff
@@ -0,0 +1,661 @@
+			}
+			xs = append(xs, key_value)
+		case zQuote:
+			return &ParseError{"", "svc key can't contain double quotes", l}
+		case zBlank:
+		default:
+			return &ParseError{"", "bad svc Values", l}
+		}
+		l, _ = c.Next()
+	}
+	rr.Value = xs
+	if rr.Priority == 0 && len(xs) > 0 {
+		return &ParseError{"", "svc aliasform can't have values", l}
+	}
+	return nil
+}
+
+// makeSvcKeyValue returns a SvcKeyValue with the key
+// or nil if a reserved key is used.
+func makeSvcKeyValue(code uint16) SvcKeyValue {
+	switch code {
+	case SVCALPN:
+		return new(SVC_ALPN)
+	case SVCNO_DEFAULT_ALPN:
+		return new(SVC_NO_DEFAULT_ALPN)
+	case SVCPORT:
+		return new(SVC_PORT)
+	case SVCIPV4HINT:
+		return new(SVC_IPV4HINT)
+	case SVCESNICONFIG:
+		return new(SVC_ESNICONFIG)
+	case SVCIPV6HINT:
+		return new(SVC_IPV6HINT)
+	default:
+		if code == 0 || code == 65535 {
+			return nil
+		}
+		e := new(SVC_LOCAL)
+		e.Code = code
+		return e
+	}
+}
+
+// SVCB RR. TODO See RFC xxxx (https://tools.ietf.org/html/draft-ietf-dnsop-svcb-httpssvc-02)
+// The one with smallest priority SHOULD be given preference.
+// Of those with equal priority, a random one SHOULD be preferred for load balancing.
+type SVCB struct {
+	Hdr      RR_Header
+	Priority uint16
+	Target   string        `dns:"domain-name"`
+	Value    []SvcKeyValue `dns:"svc"` // if priority == 0 this is empty
+}
+
+// HTTPSSVCB RR. See the beginning of this file
+type HTTPSSVC struct {
+	SVCB
+}
+
+func (rr *HTTPSSVC) String() string {
+	return rr.SVCB.String()
+}
+
+func (rr *HTTPSSVC) parse(c *zlexer, o string) *ParseError {
+	return rr.SVCB.parse(c, o)
+}
+
+func (r1 *HTTPSSVC) isDuplicate(r2 RR) bool {
+	return (*r1).SVCB.isDuplicate(r2)
+}
+
+// SvcKeyValue defines a key=value pair for SvcFieldValue.
+// A SVCB RR can have multiple SvcKeyValues appended to it.
+type SvcKeyValue interface {
+	// Key returns the key code of the pair.
+	Key() uint16
+	// pack returns the bytes of the value data.
+	pack() ([]byte, error)
+	// unpack sets the data as found in the value. Is also sets
+	// the length of the slice as the length of the value.
+	unpack([]byte) error
+	// String returns the string representation of the value.
+	String() string
+	// Read sets the data the string representation of the value.
+	Read(string) error
+	// copy returns a deep-copy of the pair.
+	copy() SvcKeyValue
+	// len returns the length of value in the wire format.
+	len() uint16
+}
+
+// SVC_ALPN pair is used to list supported connection protocols.
+// Protocol ids can be found at:
+// https://www.iana.org/assignments/tls-extensiontype-values/
+// tls-extensiontype-values.xhtml#alpn-protocol-ids
+// Basic use pattern for creating an alpn option:
+//
+//	o := new(dns.HTTPSSVC)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.TypeHTTPSSVC
+//	e := new(dns.SVC_ALPN)
+//	e.Code = dns.SVCALPN
+//	e.Alpn = []string{"h2", "http/1.1"}
+//	o.Value = append(o.Value, e)
+type SVC_ALPN struct {
+	Code uint16   // Always SVCALPN
```

Comment:
So Code isn't really need here? Maybe that can be the interface `Code() uint` ? Also the naming of the types should be move Go like and not uppercase with underscores (another mistake in the EDNS0 code). I'm not sure what would make sense here? A subpackage we're these things live? Or prefixing them with `Svc` ? Note that the final name is not chosen yet? (is this true?) So we might be looking into renaming this in the future

Reply:
Rather than prefixing every member of each struct I'd prefer to move everything to a SVCB folder/subpackage. What do you think? The specification may be extended by much in the future.

I believe `Key()` should stay in the main interface, so that its exposed parts are
```
type SvcKeyValue interface {
	// Key returns the key code of the pair.
	Key() uint16
	// String returns the string representation of the value.
	String() string
}
```